Add ListAll for HostDB httpui endpoint

bgaff opened the pull request:
We would like a way to enumerate all records in hostdb (this doesn't include the hostsfile map, although adding that would be a trivial change). I really would like feedback and suggestions.

jpeach commented:
I'm strongly in favor of introspecting HostDB, but I don't like extending HTTPUI. The right way to do this IMHO is to add management APIs so you can expose it through multiple layers (e.g. traffic_ctl). You can also add lookup and purge operations as well as list/dump.

zwoop commented:
Absolutely agree, lets not put more stuff into httpui, we should aim to eliminate it for 7.0.0. Move things over to the management port and traffic_ctl please.

jacksontj commented:
Why do we want to kill the httpui? We use it all the time for stats collection

jpeach commented:
Because generating raw HTML in core code is unmaintainable. The right approach is to extract the data and publish it in the appropriate format for the consumer (e.g. JSON or whatever).

jacksontj commented:
@jpeach from our conversations on IRC it sounds like you aren't opposed to adding this, but rather want to start making this more general. If we could make a ticket for the generic version we could start adding in all of the endpoints that we use from these and migrate them to a more unified API.

jacksontj commented:
@SolidWallOfCode How does this PR look to you?

SolidWallOfCode commented on iocore/hostdb/HostDB.cc:
```diff
@@ -1037,6 +1037,33 @@ HostDBProcessor::getbyname_imm(Continuation *cont, process_hostdb_info_pfn proce
 }
 
+Action *
+HostDBProcessor::getall(Continuation *cont)
```
Bad name. This should be "iterate" because that's what it is doing or "probeAll" to be consistent with the event / method name.

bgaff replied:
Sure, name change incoming.

SolidWallOfCode commented on iocore/hostdb/HostDB.cc:
```diff
@@ -1801,6 +1828,39 @@ HostDBContinuation::do_put_response(ClusterMachine *m, HostDBInfo *r, Continuati
+       }
+     }
+   }
+
```
Can you considered doing 1 or some small finite number of buckets per event? Otherwise you're blocking a thread for potentially a long time.

bgaff replied:
I agree, I'll add rescheduling after iterating a few buckets.

bgaff commented:
All, I've incorporated @SolidWallOfCode 's feedback which renames the method to iterate / iterateEvent, additionally i've modified the code to reschedule if it's unable to get a bucket's mutex and after 100 buckets (my local build had around 3700) so that seemed like a reasonable number, the reschedule interval for normal events is 5ms, if you can't get the mutex it uses the current 20ms timeout. Any other comments before I merge?

bgaff commented:
@zwoop This was merged but for some reason it's show as still open.
